pkg/ingester/ingester.go

```diff
@@ -241,33 +241,36 @@ func (i *Ingester) StopIncomingRequests() {
 // Push implements client.IngesterServer
 func (i *Ingester) Push(ctx old_ctx.Context, req *client.WriteRequest) (*client.WriteResponse, error) {
 	var lastPartialErr error
-	samples := client.FromWriteRequest(req)
 
-	for j := range samples {
-		err := i.append(ctx, &samples[j], req.Source)
-		if err == nil {
-			continue
-		}
-
-		ingestedSamplesFail.Inc()
-		if httpResp, ok := httpgrpc.HTTPResponseFromError(err); ok {
-			switch httpResp.Code {
-			case http.StatusBadRequest, http.StatusTooManyRequests:
-				lastPartialErr = err
+	for _, ts := range req.Timeseries {
+		for _, s := range ts.Samples {
+			err := i.append(ctx, ts.Labels, model.Time(s.TimestampMs), model.SampleValue(s.Value), req.Source)
+			if err == nil {
 				continue
 			}
-		}
 
-		return nil, err
+			ingestedSamplesFail.Inc()
+			if httpResp, ok := httpgrpc.HTTPResponseFromError(err); ok {
+				switch httpResp.Code {
+				case http.StatusBadRequest, http.StatusTooManyRequests:
+					lastPartialErr = err
+					continue
+				}
+			}
+
+			return nil, err
+		}
 	}
 
 	return &client.WriteResponse{}, lastPartialErr
 }
 
-func (i *Ingester) append(ctx context.Context, sample *model.Sample, source client.WriteRequest_SourceEnum) error {
-	for ln, lv := range sample.Metric {
-		if len(lv) == 0 {
-			delete(sample.Metric, ln)
+type labelPairs []client.LabelPair
```

Comment:
Minor nit, but this type seems dislocated; perhaps it would be better to put it next to sortedLabelPairs with a comment saying this type is unsorted?

Reply:
moved `labelPairs`